File: 2914-find-the-safest-path-in-a-grid/find-the-safest-path-in-a-grid.java
import java.util.*;

class Solution {
    // 4-directional movement array (Down, Right, Up, Left)
    static final int[][] dirs = {{1, 0}, {0, 1}, {-1, 0}, {0, -1}};

    public int maximumSafenessFactor(List<List<Integer>> grid) {
        int n = grid.size();

        // If the start or end cell contains a thief, the safeness factor is instantly 0
        if (grid.get(0).get(0) == 1 || grid.get(n - 1).get(n - 1) == 1)
            return 0;

        // Step 1: Copy the List interface matrix into a fast 2D primitive array
        int[][] A = new int[n][n];
        for (int i = 0; i < n; i++) {
            for (int j = 0; j < n; j++) {
                A[i][j] = grid.get(i).get(j);
            }
        }

        // Queue used for Multi-source Breadth-First Search (BFS)
        Queue<int[]> q = new LinkedList<>();

        // Collect all coordinates containing a thief to begin multi-source BFS
        for (int i = 0; i < n; i++) {
            for (int j = 0; j < n; j++) {
                if (A[i][j] == 1) {
                    q.add(new int[]{i, j});
                }
            }
        }

        // Step 2: Multi-source BFS to calculate the Manhattan distance to any thief.
        // It updates empty cells (0) with distances starting incrementally from 1 (1, 2, 3...)
        while (q.size() > 0) {
            int[] head = q.poll();
            int i = head[0];
            int j = head[1];
            int v = A[i][j];

            for (int[] d : dirs) {
                int x = i + d[0];
                int y = j + d[1];

                // Boundary check & confirming cell is unvisited (value is 0)
                if (Math.min(x, y) >= 0 && Math.max(x, y) < n && A[x][y] == 0) {
                    A[x][y] = v + 1; // Mark cell distance
                    q.add(new int[]{x, y});
                }
            }
        }

        // Step 3: Dijkstra-like Pathfinding via a Max-Heap (Priority Queue)
        // Ordered descending by safeness factor (b[0] - a[0]) to always explore the safest paths first
        PriorityQueue<int[]> pq = new PriorityQueue<>((a, b) -> b[0] - a[0]);
        pq.add(new int[]{A[0][0], 0, 0}); // {safeness_factor, row, col}

        while (pq.size() > 0) {
            int[] head = pq.poll();
            int sf = head[0];
            int i = head[1];
            int j = head[2];

            // If the target (bottom-right cell) is reached, return the maximum achievable safeness.
            // We subtract 1 because our BFS calculated distances starting from 1 instead of 0.
            if (i == n - 1 && j == n - 1)
                return sf - 1;

            for (int[] d : dirs) {
                int x = i + d[0];
                int y = j + d[1];

                // Boundary check & verifying cell is unvisited (positive distance value)
                if (Math.min(x, y) >= 0 && Math.max(x, y) < n && A[x][y] > 0) {
                    // Path safeness is limited by the bottleneck (minimum) cell value along the path
                    pq.add(new int[]{Math.min(sf, A[x][y]), x, y});
                    
                    // In-place tracking: Negate the value to mark it visited without using an extra boolean array
                    A[x][y] *= -1;
                }
            }
        }

        return A[n - 1][n - 1] - 1;
    }
}

File: 2914-find-the-safest-path-in-a-grid/find-the-safest-path-in-a-grid.py
from collections import deque
import heapq

class Solution:
    def maximumSafenessFactor(self, grid: list[list[int]]) -> int:
        n = len(grid)
        
        # If the start or end cell has a thief, safeness factor is immediately 0
        if grid[0][0] == 1 or grid[n-1][n-1] == 1:
            return 0
            
        # Initialize the safeness distance grid with infinity
        # dist[r][c] will store the Manhattan distance to the nearest thief
        dist = [[float('inf')] * n for _ in range(n)]
        queue = deque()
        
        # Step 1: Add all thieves to the queue and set their distance to 0
        for r in range(n):
            for c in range(n):
                if grid[r][c] == 1:
                    dist[r][c] = 0
                    queue.append((r, c))
                    
        # Multi-source BFS to calculate the shortest distance from each cell to any thief
        directions = [(0, 1), (0, -1), (1, 0), (-1, 0)]
        while queue:
            r, c = queue.popleft()
            for dr, dc in directions:
                nr, nc = r + dr, c + dc
                # If neighbor is valid and we found a shorter path to a thief
                if 0 <= nr < n and 0 <= nc < n and dist[nr][nc] == float('inf'):
                    dist[nr][nc] = dist[r][c] + 1
                    queue.append((nr, nc))
                    
        # Step 2: Find the maximum safeness path using a Max-Heap (Dijkstra-like)
        # Python's heapq is a min-heap by default, so we store negative values to simulate a max-heap
        max_heap = [(-dist[0][0], 0, 0)]  # (negative safeness, row, col)
        
        # Track maximum safeness encountered so far for each cell to avoid redundant processing
        max_safeness = [[-1] * n for _ in range(n)]
        max_safeness[0][0] = dist[0][0]
        
        while max_heap:
            d, r, c = heapq.heappop(max_heap)
            current_safeness = -d
            
            # If we reached the bottom-right cell, return the maximum safeness factor achieved
            if r == n - 1 and c == n - 1:
                return current_safeness
                
            for dr, dc in directions:
                nr, nc = r + dr, c + dc
                if 0 <= nr < n and 0 <= nc < n:
                    # The safeness of the path to the neighbor is limited by 
                    # either the current path's safeness or the neighbor's own safeness value
                    path_safeness = min(current_safeness, dist[nr][nc])
                    
                    # If this path offers a higher safeness factor than previously found for this cell
                    if path_safeness > max_safeness[nr][nc]:
                        max_safeness[nr][nc] = path_safeness
                        heapq.heappush(max_heap, (-path_safeness, nr, nc))
                        
        return 0